Python/compile.c

```diff
@@ -701,6 +702,19 @@ compiler_set_qualname(struct compiler *c)
         capsule = PyList_GET_ITEM(c->c_stack, stack_size - 1);
         parent = (struct compiler_unit *)PyCapsule_GetPointer(capsule, CAPSULE_NAME);
         assert(parent);
+        if (parent->u_scope_type == COMPILER_SCOPE_TYPEPARAMS) {
```

Comment:
Would it be possible to just set the qualname of a COMPILER_SCOPE_TYPEPARAMS codeunit to be the same as the qualname of its parent, and then avoid all the special handling here?

Reply:
I don't think so, because we'd still insert the `<locals>` part into the qualname. If you have
```python
def f[T](): pass
```
Technically `f` is now defined within the locals of an enclosing function, so we need some special handling here to get rid of the `<locals>` part.

Reply:
You're right, tests pass if I just remove this block. I think I wrote this before there was a separate COMPILER_SCOPE for type params.

Reply:
Sorry, that's not true; I think I got confused with old pyc files when I tried this.

Setting the qualname of the TYPEPARAMS scope to the qualname of the parent fails at the top level, because the module scope itself has no qualname, so the qualname for top-level generic functions would be set incorrectly. There are other ways around that, but I'm not sure they're simpler than the code I previously had here.